lock-api/src/test/java/com/palantir/lock/client/MultiClientTransactionStarterTest.java

```diff
@@ -0,0 +1,211 @@
+                true);
+    }
+
+    @Test
+    public void canServiceMultipleClientsWithMultipleServerCalls() {
+        int clientCount = 5;
+        assertSanityOfResponse(
+                getStartTransactionRequestsForClients(
+                        clientCount, (PARTITIONED_TIMESTAMPS_LIMIT_PER_SERVER_CALL + 1) * clientCount),
+                true);
+    }
+
+    @Test
+    public void canServiceOneRequestWithMultipleServerRequests() {
+        Namespace namespace = Namespace.of("Test_0");
+        assertSanityOfResponse(
+                ImmutableList.of(BatchElement.of(
+                        NamespaceAndRequestParams.of(
+                                namespace, RequestParams.of(127, getCache(namespace), lockCleanupService)),
+                        new DisruptorFuture<>("test"))),
+                false);
+    }
+
+    @Test
+    public void updatesCacheWhileProcessingResponse() {
+        Namespace namespace = Namespace.of("Test" + UUID.randomUUID());
+        assertSanityOfResponse(
+                ImmutableList.of(BatchElement.of(
+                        NamespaceAndRequestParams.of(
+                                namespace,
+                                RequestParams.of(
+                                        PARTITIONED_TIMESTAMPS_LIMIT_PER_SERVER_CALL - 1,
+                                        getCache(namespace),
+                                        lockCleanupService)),
+                        new DisruptorFuture<>("test"))),
+                true);
+        verify(getCache(namespace)).processStartTransactionsUpdate(any(), any());
+    }
+
+    @Test
+    public void shouldFreeResourcesIfServerThrows() {
+        Namespace namespace = Namespace.of("Test" + UUID.randomUUID());
+
+        UUID requestorId = UUID.randomUUID();
+        ImmutableList<BatchElement<NamespaceAndRequestParams, List<StartIdentifiedAtlasDbTransactionResponse>>>
+                requests = ImmutableList.of(BatchElement.of(
+                NamespaceAndRequestParams.of(
+                        namespace,
+                        RequestParams.of(
+                                PARTITIONED_TIMESTAMPS_LIMIT_PER_SERVER_CALL * 5,
+                                getCache(namespace),
+                                lockCleanupService)),
+                new DisruptorFuture<>("test")));
+        Map<Namespace, ConjureStartTransactionsResponse> responseMap =
+                getMultiClientStartTransactionsResponse(requests, requestorId);
+
+        SafeIllegalStateException exception = new SafeIllegalStateException("Something went wrong!");
+        when(timelockService.startTransactions(any())).thenReturn(responseMap).thenThrow(exception);
+
+        assertThatThrownBy(() -> processBatch(timelockService, requestorId, requests))
+                .isEqualTo(exception);
+        verify(lockCleanupService).refreshLockLeases(any());
+        verify(lockCleanupService).unlock(any());
+    }
+
```

Comment:
I think we need one more test, which unfortunately may not be that easy to write (I'm fine whether this is separate or rolled into the above one): the lock tokens corresponding to requests that were already served are _not_ automatically unlocked.

You will likely need to use `ArgumentCaptor` or some mechanism for inspecting what exactly was unlocked in this case.

Reply:
we paired on this! turns out the argumentcaptor can do some pretty cool stuff here